src/compiler.rkt

```diff
@@ -223,108 +219,106 @@
-                      (- (max xhi-exp yhi-exp) outhi-exp)))]))]
+       (max 0 (+ 1 ; subtraction of logarithms doesn't consider mantissa - which can be +1 to the result
+                 (max
+                  (- (max xlo-exp ylo-exp) outlo-exp)
+                  (- (max xhi-exp yhi-exp) outhi-exp))
+                 (if (and (or (not (equal? xlo-sgn ylo-sgn)) ; slack part
+                              (not (equal? xhi-sgn yhi-sgn)))
+                          (or (>= 2 (abs (- xlo-exp ylo-exp)))
+                              (>= 2 (abs (- xhi-exp yhi-exp)))))
+                     (get-slack)
+                     0)))]
       
       [(equal? op ival-sub)
        ; log[Г-] = max(log[x], log[y]) - log[x - y]
+       ;                               ^^^^^^^^^^^^
+       ;                               possible cancellation
        (define x (first srcs))
        (define xlo (ival-lo x))
-       (define xlo-exp (true-exponent xlo))
+       (define xlo-exp (log2-approx xlo))
        (define xlo-sgn (bigfloat-signbit xlo))
        (define xhi (ival-hi x))
-       (define xhi-exp (true-exponent xhi))
+       (define xhi-exp (log2-approx xhi))
        (define xhi-sgn (bigfloat-signbit xhi))
 
        (define y (second srcs))
        (define ylo (ival-lo y))
-       (define ylo-exp (true-exponent ylo))
+       (define ylo-exp (log2-approx ylo))
        (define ylo-sgn (bigfloat-signbit ylo))
        (define yhi (ival-hi y))
-       (define yhi-exp (true-exponent yhi))
+       (define yhi-exp (log2-approx yhi))
        (define yhi-sgn (bigfloat-signbit yhi))
 
        (define outlo (ival-lo output))
-       (define outlo-exp (true-exponent outlo))
+       (define outlo-exp (log2-approx outlo))
        (define outhi (ival-hi output))
-       (define outhi-exp (true-exponent outhi))
+       (define outhi-exp (log2-approx outhi))
 
-       (max 0
-            (match (and (or (equal? xlo-sgn yhi-sgn)
-                            (equal? xhi-sgn ylo-sgn))
-                        (or (>= 2 (abs (- xlo-exp yhi-exp)))
-                            (>= 2 (abs (- xhi-exp ylo-exp)))))
-              [#f (+ 1
-                     (max
-                      (- (max xlo-exp yhi-exp) outlo-exp)
-                      (- (max xhi-exp ylo-exp) outhi-exp)))]
-              [#t (+ (get-slack)
-                     (max
-                      (- (max xlo-exp yhi-exp) outlo-exp)
-                      (- (max xhi-exp ylo-exp) outhi-exp)))]))]
+       (max 0 (+ 1 ; subtraction of logarithms doesn't consider mantissa - which can be +1 to the result
+                 (max
+                  (- (max xlo-exp yhi-exp) outlo-exp)
+                  (- (max xhi-exp ylo-exp) outhi-exp))
+                 (if (and (or (equal? xlo-sgn yhi-sgn) ; slack part
+                              (equal? xhi-sgn ylo-sgn))
+                          (or (>= 2 (abs (- xlo-exp yhi-exp)))
+                              (>= 2 (abs (- xhi-exp ylo-exp)))))
```

Comment:
Why does this look only at the inputs? Can't we just look at whether the output interval crosses zero? Here you're basically guessing whether or not it will.

Reply:
I tried to leave this solution for nightly, here is the repository of how it looked like: [tree](https://github.com/herbie-fp/herbie/tree/864d0f2d9771de8791b6edd8e8f1d5a94db399fa)
And here is nightly - exit points raised to 300k and benchmarks that used to converge didn't work quite well
[report](https://nightly.cs.washington.edu/reports/herbie/1711277705:nightly:artem-minus-and-trig-tuning:864d0f2d97/)

Reply:
Everything seems to work good in theory but smth went wrong..

Reply:
Huh. Ok, I mean, something to figure out later but it shouldn't interfere with this PR.